### PR TITLE
[ao] fixing tutorials

### PR DESCRIPTION
**Pull request by HDCharles**
Summary: fixing tutorial to have example_inputs and QConfigMapping.

note: the quantization guide leaves example_inputs non initialized that code doesn't seem intended to run and we don't know what teh model is supposed to be.

Test Plan: for fx_graph_mode_ptq_static.rst and
fx_graph_mode_ptq_dynamic.py I pasted and ran the script to verify that it ran, the latter could be run exactly while the former required a slight modification since I didn't had a different version of the imagenet dataset but only the dataloaders had to be modified.

for the quant_guide no explicit testing was done
Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/tutorials/pull/2137

### CODE REVIEW

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 1]
```diff
@@ -1,64 +1,66 @@
...
+-------------------------------------------
+
+Currently PyTorch only has eager mode quantization: `Static Quantization with Eager Mode in PyTorch <https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html>`_.
+
+We can see there are multiple manual steps involved in the process, including:
+
+- Explicitly quantize and dequantize activations, this is time consuming when floating point and quantized operations are mixed in a model.
+- Explicitly fuse modules, this requires manually identifying the sequence of convolutions, batch norms and relus and other fusion patterns.
+- Special handling is needed for pytorch tensor operations (like add, concat etc.)
+- Functionals did not have first class support (functional.conv2d and functional.linear would not get quantized)
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:44]:
```suggestion
- Functionals did not have first class support—``functional.conv2d`` and ``functional.linear`` would not get quantized)
```

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 2]
```diff
@@ -1,64 +1,66 @@
...
+
+Most of these required modifications comes from the underlying limitations of eager mode quantization. Eager mode works in module level since it can not inspect the code that is actually run (in the forward function), quantization is achieved by module swapping, and we don’t know how the modules are used in forward function in eager mode, so it requires users to insert QuantStub and DeQuantStub manually to mark the points they want to quantize or dequantize.
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:46]:
```suggestion
Most of these required modifications come from the underlying limitations of eager mode quantization. Eager mode works in module level since it can not inspect the code that is actually run (in the forward function), quantization is achieved by module swapping, and we don’t know how the modules are used in forward function in eager mode, so it requires users to insert QuantStub and DeQuantStub manually to mark the points they want to quantize or dequantize.
```

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 3]
```diff
@@ -1,64 +1,66 @@
...
+In graph mode, we can inspect the actual code that’s been executed in forward function (e.g. aten function calls) and quantization is achieved by module and graph manipulations. Since graph mode has full visibility of the code that is run, our tool is able to automatically figure out things like which modules to fuse and where to insert observer calls, quantize/dequantize functions etc., we are able to automate the whole quantization process.
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:47]:
```suggestion
In graph mode, we can inspect the actual code that’s been executed in forward function (for example, aten function calls) and quantization is achieved by module and graph manipulations. Since graph mode has full visibility of the code that is run, our tool is able to automatically figure out things like which modules to fuse and where to insert observer calls, quantize/dequantize functions, and so on, we are able to automate the whole quantization process.
```

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 4]
```diff
@@ -1,64 +1,66 @@
...
+
+Advantages of FX Graph Mode Quantization are:
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:49]:
```suggestion
Advantages of FX Graph Mode Quantization include:
```

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 5]
```diff
@@ -1,64 +1,66 @@
...
+
+- Simple quantization flow, minimal manual steps
+- Unlocks the possibility of doing higher level optimizations like automatic precision selection
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:52]:
```suggestion
- The possibility of performing higher level optimizations like automatic precision selection
```

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 6]
```diff
@@ -185,252 +187,223 @@ Download the `torchvision resnet18 model <https://download.pytorch.org/models/re
...
+      .set_module_name("foo.bar", qconfig?)
+      .set_module_name_object_type_order()
+  )
+      # priority (in increasing order): global, object_type, module_name_regex, module_name
+      # qconfig == None means fusion and quantization should be skipped for anything
+      # matching the rule (unless a higher priority match is found)
+
+
+Utility functions related to ``qconfig`` can be found in the `qconfig <https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/qconfig.py>`_ file
+while those for ``QConfigMapping`` can be found in the `qconfig_mapping <https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/qconfig_mapping.py>`
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:255]:
```suggestion
while those for ``QConfigMapping`` can be found in the `qconfig_mapping <https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/qconfig_mapping.py>`__
```

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 7]
```diff
@@ -185,252 +187,223 @@ Download the `torchvision resnet18 model <https://download.pytorch.org/models/re
...
-    print(prepared_model.graph) 
+    prepared_model = prepare_fx(model_to_quantize, qconfig_mapping, example_inputs)
 
-6. Calibration    
---------------    
-Calibration function is run after the observers are inserted in the model.    
-The purpose for calibration is to run through some sample examples that is representative of the workload     
-(for example a sample of the training data set) so that the observers in the model are able to observe    
-the statistics of the Tensors and we can later use this information to calculate quantization parameters. 
+prepare_fx folds BatchNorm modules into previous Conv2d modules, and insert observers
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:269]:
```suggestion
``prepare_fx`` folds BatchNorm modules into previous Conv2d modules, and insert observers
```

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 8]
```diff
@@ -185,252 +187,223 @@ Download the `torchvision resnet18 model <https://download.pytorch.org/models/re
...
-    calibrate(prepared_model, data_loader_test)  # run calibration on sample data   
+    prepared_model = prepare_fx(model_to_quantize, qconfig_mapping, example_inputs)
+    print(prepared_model.graph)
+
+6. Calibration
+--------------
+Calibration function is run after the observers are inserted in the model.
+The purpose for calibration is to run through some sample examples that is representative of the workload
+(for example a sample of the training data set) so that the observers in the model are able to observe
+the statistics of the Tensors and we can later use this information to calculate quantization parameters.
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:282]:
```suggestion
the statistics of the Tensors, and we can later use this information to calculate quantization parameters.
```

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 9]
```diff
@@ -185,252 +187,223 @@ Download the `torchvision resnet18 model <https://download.pytorch.org/models/re
...
+    # loaded_quantized_model.load_state_dict(torch.load(fx_graph_mode_model_file_path))
+
+    # save with script
+    torch.jit.save(torch.jit.script(quantized_model), fx_graph_mode_model_file_path)
+    loaded_quantized_model = torch.jit.load(fx_graph_mode_model_file_path)
+
+    top1, top5 = evaluate(loaded_quantized_model, criterion, data_loader_test)
+    print("[after serialization/deserialization] Evaluation accuracy on test dataset: %2.2f, %2.2f"%(top1.avg, top5.avg))
+
+If you want to get better accuracy or performance,  try changing the `qconfig_mapping`.
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:338]:
```suggestion
If you want to get better accuracy or performance, try changing the ``qconfig_mapping``.
```

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 10]
```diff
@@ -185,252 +187,223 @@ Download the `torchvision resnet18 model <https://download.pytorch.org/models/re
...
+We plan to add support for graph mode in the Numerical Suite so that you can
+easily determine the sensitivity towards quantization of different modules in a model: `PyTorch Numeric Suite Tutorial <https://pytorch.org/tutorials/prototype/numeric_suite_tutorial.html>`_
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:340]:
```suggestion
easily determine the sensitivity towards quantization of different modules in a model. For more information, see `PyTorch Numeric Suite Tutorial <https://pytorch.org/tutorials/prototype/numeric_suite_tutorial.html>`_
```

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 11]
```diff
@@ -185,252 +187,223 @@ Download the `torchvision resnet18 model <https://download.pytorch.org/models/re
...
+
+9. Debugging Quantized Model
+----------------------------
+We can also print the weight for quantized an un-quantized conv to see the difference,
+we'll first call fuse explicitly to fuse the conv and bn in the model:
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:345]: should it be: "to fuse the convolution and batch norm layers in the model"

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 12]
```diff
@@ -185,252 +187,223 @@ Download the `torchvision resnet18 model <https://download.pytorch.org/models/re
...
-    torch.jit.save(torch.jit.script(float_model), saved_model_dir + scripted_float_model_file)  
-  
-In this section we compare the model quantized with FX graph mode quantization with the model     
-quantized in eager mode. FX graph mode and eager mode produce very similar quantized models,  
-so the expectation is that the accuracy and speedup are similar as well.  
+    top1, top5 = evaluate(float_model, criterion, data_loader_test)
+    print("Baseline Float Model Evaluation accuracy: %2.2f, %2.2f"%(top1.avg, top5.avg))
+    torch.jit.save(torch.jit.script(float_model), saved_model_dir + scripted_float_model_file)
+
+In this section we compare the model quantized with FX graph mode quantization with the model
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:371]:
```suggestion
In this section, we compare the model quantized with FX graph mode quantization with the model
```

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 13]
```diff
@@ -185,252 +187,223 @@ Download the `torchvision resnet18 model <https://download.pytorch.org/models/re
...
+  Scripted Float Model:
+  Self CPU time total: 192.48ms
+
+  Scripted Eager Mode Quantized Model:
+  Self CPU time total: 50.76ms
+
+  Scripted FX Graph Mode Quantized Model:
+  Self CPU time total: 50.63ms
+
+As we can see for resnet18 both FX graph mode and eager mode quantized model get similar speed up over the floating point model,
```
Comment (svekars) [inline prototype_source/fx_graph_mode_ptq_static.rst:407]:
```suggestion
As we can see for resnet18 both FX graph mode and eager mode quantized model get similar speedup over the floating point model,
```

File: prototype_source/fx_graph_mode_quant_guide.rst [hunk 14]
```diff
@@ -3,13 +3,13 @@
...
+FX Graph Mode Quantization requires a symbolically traceable model.
 We use the FX framework (TODO: link) to convert a symbolically traceable nn.Module instance to IR,
-and we operate on the IR to execute the quantization passes.  
+and we operate on the IR to execute the quantization passes.
 Please post your question about symbolically tracing your model in `PyTorch Discussion Forum <https://discuss.pytorch.org/c/quantization/17>`_
 
-Quantization will only work on the symbolically traceable parts of your model. 
-Data dependent control flow (if statements / for loops etc using symbolically traced values) are one common pattern which is not supported. 
+Quantization will only work on the symbolically traceable parts of your model.
+Data dependent control flow (if statements / for loops etc using symbolically traced values) are one common pattern which is not supported.
```
Comment (svekars) [inline prototype_source/fx_graph_mode_quant_guide.rst:12]:
```suggestion
The data dependent control flow—if statements / for loop, and so on using symbolically traced values—are one common pattern which is not supported.
```

**Review by svekars — COMMENTED** [review]
Looks good! Just a couple editorial suggestions.

File: prototype_source/fx_graph_mode_ptq_static.rst [hunk 15]
```diff
@@ -1,64 +1,66 @@
...
-although there might some effort required to make the model compatible with FX Graph Mode Quantizatiion (symbolically traceable with ``torch.fx``), 
-we'll have a separate tutorial to show how to make the part of the model we want to quantize compatibble with FX Graph Mode Quantization.   
+(prototype) FX Graph Mode Post Training Static Quantization
+===========================================================
+**Author**: `Jerry Zhang <https://github.com/jerryzh168>`_ **Edited by**: `Charles Hernandez <https://github.com/HDCharles>`_
+
+This tutorial introduces the steps to do post training static quantization in graph mode based on
+`torch.fx <https://github.com/pytorch/pytorch/blob/master/torch/fx/__init__.py>`_.
+The advantage of FX graph mode quantization is that we can perform quantization fully automatically on the model
+although there might some effort required to make the model compatible with FX Graph Mode Quantizatiion (symbolically traceable with ``torch.fx``),
```
[the remark this answers is not in the document]
Comment (HDCharles) [reply prototype_source/fx_graph_mode_ptq_static.rst]: good catch

**Comment by HDCharles** [comment]
thanks for your help cleaning up these tutorials @svekars, i've made those suggested changes

**Review by svekars — APPROVED** [review]
LGTM!
